core/consensus/aura/src/lib.rs

```diff
@@ -379,80 +298,126 @@ impl<B: Block, C, E, I, P, Error, SO> SlotWorker<B> for AuraWorker<C, E, I, P, S
 				// deadline our production to approx. the end of the
 				// slot
 				Timeout::new(
-					proposer.propose(slot_info.inherent_data, remaining_duration).into_future(),
+					proposer.propose(
+						slot_info.inherent_data,
+						remaining_duration,
+						generic::Digest {
+							logs: vec![
+								<DigestItemFor<B> as CompatibleDigestItem<P>>::aura_pre_digest(slot_num),
+							],
+						},
+					).into_future(),
 					remaining_duration,
 				)
 			} else {
 				return Box::new(future::ok(()));
 			}
 		};
 
-		Box::new(
-			proposal_work
-				.map(move |b| {
-					// minor hack since we don't have access to the timestamp
-					// that is actually set by the proposer.
-					let slot_after_building = slot_now(slot_duration);
-					if slot_after_building != Some(slot_num) {
-						info!(
-							"Discarding proposal for slot {}; block production took too long",
-							slot_num
-						);
-						telemetry!(CONSENSUS_INFO; "aura.discarding_proposal_took_too_long";
-							"slot" => slot_num
-						);
-						return
-					}
+		Box::new(proposal_work.map(move |b| {
+			// minor hack since we don't have access to the timestamp
+			// that is actually set by the proposer.
+			let slot_after_building = slot_now(slot_duration);
+			if slot_after_building != Some(slot_num) {
+				info!(
+					"Discarding proposal for slot {}; block production took too long",
+					slot_num
+				);
+				telemetry!(CONSENSUS_INFO; "aura.discarding_proposal_took_too_long";
+					"slot" => slot_num
+				);
+				return
+			}
 
-					let (header, body) = b.deconstruct();
-					let header_num = header.number().clone();
-					let pre_hash = header.hash();
-					let parent_hash = header.parent_hash().clone();
+			let (mut header, body) = b.deconstruct();
+			// TODO is better debug logging worth hashing the header again?
+			let header_hash = header.hash();
+			let pre_digest: Result<u64, String> = find_pre_digest::<B, P>(&header, header_hash);
+			if pre_digest.is_err() {
+				warn!(target: "aura", "Runtime stripped our digest!  Re-adding it.");
+				header.append_digest(generic::Digest {
+					logs: vec![
+						generic::DigestItem::aura_pre_digest(slot_num)
+					],
+				})
+			} else {
+				trace!(target: "aura", "Got correct number of seals.  Good!")
+			};
+			let header_num = header.number().clone();
+			let pre_hash = header.hash();
```

Comment:
if my above comment is right and we never have to re-push pre-digests, this is duplicated with the `header_hash` value.